lib/fs/basicfs_copy_range_copyfilerange.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// +build linux
+
+package fs
+
+import (
+	"io"
+	"syscall"
+
+	"golang.org/x/sys/unix"
+)
+
+func init() {
+	registerCopyRangeImplementation(CopyRangeMethodCopyFileRange, copyRangeImplementationForBasicFile(copyRangeCopyFileRange))
+}
+
+func copyRangeCopyFileRange(src, dst basicFile, srcOffset, dstOffset, size int64) error {
+	for size > 0 {
+		// From MAN page:
+		//
+		// If off_in is not NULL, then off_in must point to a buffer that
+		// specifies the starting offset where bytes from fd_in will be read.
+		// 	The file offset of fd_in is not changed, but off_in is adjusted
+		// appropriately.
+		//
+		// Also, even if explicitly not stated, the same is true for dstOffset
+		n, err := unix.CopyFileRange(int(src.Fd()), &srcOffset, int(dst.Fd()), &dstOffset, int(size), 0)
+		if n == 0 && err == nil {
+			return io.ErrUnexpectedEOF
+		}
+		if err != nil && err != syscall.EAGAIN {
+			return err
+		}
+		size -= int64(n)
```

Comment:
if we got syscall.EAGAIN above we will happily reduce size and continue?

Reply:
EAGAIN usually means interrupted, which could copy some data... We're deducting n, so we'll know how much, if anything, was copied.

Reply:
Ah, cool.

Reply:
Looking at:
https://go-review.googlesource.com/c/go/+/229101/7/src/internal/poll/copy_file_range_linux.go

It seems it does not need any special handling.
Also, it seems io.Copy now does this behind the scenes.

Reply:
Apparently there is no such thing as non-blocking file io, so there is no way that EAGAIN could be returned.

Reply:
Well that was the theory about Stat() as well and that doesn’t match up with reality so I think handling this is good; unless failure results in fallback to userspace copies.

Reply:
So it's not clear what's the value for N when EAGAIN returns.  
The docs state if n is -1, check errno, implying that if err == EAGAIN, then n = -1, so I am going for the safe option and not handling it.

I guess I could do
```
if err != nil && err != EAGAIN {
}
if n > 0 {
  size -= n
}
```

Reply:
Updated to deduct n only if it's positive.